File: nhglib-core/src/io/github/voidzombie/nhglib/data/models/serialization/physics/shapes/ShapeJson.java
package io.github.voidzombie.nhglib.data.models.serialization.physics.shapes;

import com.badlogic.gdx.utils.JsonValue;
import io.github.voidzombie.nhglib.interfaces.JsonParseable;
import io.github.voidzombie.nhglib.physics.enums.RigidBodyType;
import io.github.voidzombie.nhglib.physics.models.RigidBodyShape;

/**
 * Created by Fausto Napoli on 16/05/2017.
 */
public class ShapeJson implements JsonParseable<RigidBodyShape> {
    public RigidBodyShape output;

    @Override
    public void parse(JsonValue shapeJson) {
        RigidBodyType type = RigidBodyType.fromString(shapeJson.getString("type"));

        switch (type) {
            case SPHERE:
                SphereShapeJson sphereShapeJson = new SphereShapeJson();
                sphereShapeJson.parse(shapeJson);
                output = sphereShapeJson.get();
                break;

            case BOX:
                BoxShapeJson boxShapeJson = new BoxShapeJson();
                boxShapeJson.parse(shapeJson);
                output = boxShapeJson.get();
                break;

            case CYLINDER:
                CylinderShapeJson cylinderShapeJson = new CylinderShapeJson();
                cylinderShapeJson.parse(shapeJson);
                output = cylinderShapeJson.get();
                break;

            case CONE:
                ConeShapeJson coneShapeJson = new ConeShapeJson();
                coneShapeJson.parse(shapeJson);
                output = coneShapeJson.get();
                break;

            case CAPSULE:
                CapsuleShapeJson capsuleShapeJson = new CapsuleShapeJson();
                capsuleShapeJson.parse(shapeJson);
                output = capsuleShapeJson.get();
                break;

            case CONVEX_HULL:
                ConvexHullShapeJson convexHullShapeJson = new ConvexHullShapeJson();
                convexHullShapeJson.parse(shapeJson);
                output = convexHullShapeJson.get();
                break;

            case BVH_TRIANGLE_MESH:
                BvhTriangleMeshShapeJson bvhTriangleMeshShapeJson = new BvhTriangleMeshShapeJson();
                bvhTriangleMeshShapeJson.parse(shapeJson);
                output = bvhTriangleMeshShapeJson.get();
                break;

            case CONVEX_TRIANGLE_MESH:
                ConvexTriangleMeshShapeJson convexTriangleMeshShapeJson = new ConvexTriangleMeshShapeJson();
                convexTriangleMeshShapeJson.parse(shapeJson);
                output = convexTriangleMeshShapeJson.get();
                break;
        }

        output.type = type;
    }

    @Override
    public RigidBodyShape get() {
        return output;
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/data/models/serialization/physics/shapes/ShapeJson.kt
package io.github.movementspeed.nhglib.data.models.serialization.physics.shapes

import com.badlogic.gdx.utils.JsonValue
import io.github.movementspeed.nhglib.interfaces.JsonParseable
import io.github.movementspeed.nhglib.physics.enums.RigidBodyType
import io.github.movementspeed.nhglib.physics.models.RigidBodyShape

/**
 * Created by Fausto Napoli on 16/05/2017.
 */
open class ShapeJson : JsonParseable<RigidBodyShape> {
    var output: RigidBodyShape? = null

    override fun parse(shapeJson: JsonValue) {
        val type = RigidBodyType.fromString(shapeJson.getString("type"))

        when (type) {
            RigidBodyType.SPHERE -> {
                val sphereShapeJson = SphereShapeJson()
                sphereShapeJson.parse(shapeJson)
                output = sphereShapeJson.get()
            }

            RigidBodyType.BOX -> {
                val boxShapeJson = BoxShapeJson()
                boxShapeJson.parse(shapeJson)
                output = boxShapeJson.get()
            }

            RigidBodyType.CYLINDER -> {
                val cylinderShapeJson = CylinderShapeJson()
                cylinderShapeJson.parse(shapeJson)
                output = cylinderShapeJson.get()
            }

            RigidBodyType.CONE -> {
                val coneShapeJson = ConeShapeJson()
                coneShapeJson.parse(shapeJson)
                output = coneShapeJson.get()
            }

            RigidBodyType.CAPSULE -> {
                val capsuleShapeJson = CapsuleShapeJson()
                capsuleShapeJson.parse(shapeJson)
                output = capsuleShapeJson.get()
            }

            RigidBodyType.CONVEX_HULL -> {
                val convexHullShapeJson = ConvexHullShapeJson()
                convexHullShapeJson.parse(shapeJson)
                output = convexHullShapeJson.get()
            }

            RigidBodyType.BVH_TRIANGLE_MESH -> {
                val bvhTriangleMeshShapeJson = BvhTriangleMeshShapeJson()
                bvhTriangleMeshShapeJson.parse(shapeJson)
                output = bvhTriangleMeshShapeJson.get()
            }

            RigidBodyType.CONVEX_TRIANGLE_MESH -> {
                val convexTriangleMeshShapeJson = ConvexTriangleMeshShapeJson()
                convexTriangleMeshShapeJson.parse(shapeJson)
                output = convexTriangleMeshShapeJson.get()
            }
        }

        output!!.type = type
    }

    override fun get(): RigidBodyShape {
        return output
    }
}
